data/tests_jobsummary.py

```diff
@@ -0,0 +1,122 @@
+        self.assertEqual(pair[0].state, Job.JOB_STATE_RUNNING)  # subsequent item is RUNNING
+        self.assertEqual(pair[0].step, Job.JOB_STEP_RUNNING)  # subsequent item is RUNNING
+        self.assertEqual(pair[1], None)  # last item is None because there is no next item
+
+    def test_calculate_elapsed_hours(self):
+        hours = JobSummary._calculate_elapsed_hours(
+            activity=Mock(created=self.created_ts('12:00')),
+            next_activity=Mock(created=self.created_ts('14:30')))
+        self.assertEqual(hours, 2.5)
+
+    @patch('data.jobsummary.datetime')
+    def test_vm_hours_still_running(self, mock_datetime):
+        activities = [
+            (Job.JOB_STATE_NEW, '', self.created_ts('11:50')),
+            (Job.JOB_STATE_AUTHORIZED, '', self.created_ts('11:58')),
+            (Job.JOB_STATE_RUNNING, '', self.created_ts('11:59')),
+            (Job.JOB_STATE_RUNNING, Job.JOB_STEP_STAGING, self.created_ts('12:00')),
+            (Job.JOB_STATE_RUNNING, Job.JOB_STATE_RUNNING, self.created_ts('12:30')),
+        ]
+        mock_job = self.mock_job(activities, num_cpus=32)
+        mock_datetime.datetime.now.return_value = self.created_ts('14:30')
+        summary = JobSummary(mock_job)
+        self.assertEqual(summary.vm_hours, 2.5)
+
+    @patch('data.jobsummary.datetime')
+    def test_vm_hours_finished(self, mock_datetime):
+        activities = [
+            (Job.JOB_STATE_NEW, '', self.created_ts('11:50')),
+            (Job.JOB_STATE_AUTHORIZED, '', self.created_ts('11:58')),
+            (Job.JOB_STATE_RUNNING, '', self.created_ts('11:59')),
+            (Job.JOB_STATE_RUNNING, Job.JOB_STEP_STAGING, self.created_ts('12:00')),
+            (Job.JOB_STATE_RUNNING, Job.JOB_STATE_RUNNING, self.created_ts('12:30')),
+            (Job.JOB_STATE_RUNNING, Job.JOB_STEP_TERMINATE_VM, self.created_ts('13:15')),
+            (Job.JOB_STATE_FINISHED, '', self.created_ts('13:16')),
+        ]
+        mock_job = self.mock_job(activities, num_cpus=32)
+        summary = JobSummary(mock_job)
+        mock_datetime.datetime.now.return_value = self.created_ts('14:30')
+        self.assertEqual(summary.vm_hours, 1.25)
+
+    @patch('data.jobsummary.datetime')
+    def test_vm_hours_error(self, mock_datetime):
+        activities = [
+            (Job.JOB_STATE_NEW, '', self.created_ts('11:50')),
+            (Job.JOB_STATE_AUTHORIZED, '', self.created_ts('11:58')),
+            (Job.JOB_STATE_RUNNING, '', self.created_ts('11:59')),
+            (Job.JOB_STATE_RUNNING, Job.JOB_STEP_STAGING, self.created_ts('12:00')),
+            (Job.JOB_STATE_RUNNING, Job.JOB_STATE_RUNNING, self.created_ts('12:30')),
+            (Job.JOB_STATE_ERROR, Job.JOB_STATE_RUNNING, self.created_ts('13:15')),
+        ]
+        mock_job = self.mock_job(activities, num_cpus=32)
+        summary = JobSummary(mock_job)
+        mock_datetime.datetime.now.return_value = self.created_ts('14:30')
+        self.assertEqual(summary.vm_hours, 1.25)
```

Comment:
I don't see a test that covers the issue in #113 - if a job is restarted after an error, is the idle time between the error and the restart counted?